dix/req.c

```diff
@@ -2214,8 +2237,16 @@ static int dix_cas_rops_alloc(struct m0_dix_req *req)
 		}
 		for (j = 0; j < rec_op->dgp_units_nr; j++) {
 			unit = &rec_op->dgp_units[j];
-			if (dix_pg_unit_skip(req, unit))
+			if (dix_pg_unit_skip(req, unit)) {
+				if (dtx != NULL &&
+				    unit->dpu_pd_state == M0_PNDS_OFFLINE &&
+				    skipped_sdevs[unit->dpu_tgt] == NULL) {
+					skipped_sdevs[unit->dpu_tgt] =
+						&unit->dpu_sdev_idx;
```

Comment:
Can we define skipped_sdevs as array of uint32_t?
```
uint32_t *skipped_sdevs = NULL;
if (dtx != NULL) {
		M0_ALLOC_ARR(skipped_sdevs, skipped_sdevs_max);
		if (skipped_sdevs == NULL)
			return M0_ERR(-ENOENT);
}
for (pa_idx = 0; pa_idx < skipped_sdevs_max; skipped_sdevs_max++)
        skipped_sdevs[pa_idx] = 0xffffffff;
```
So we can use 0xfffffff as the default value to indicate it is not used.
I know now it is an array of pointer and NULL is the default value of its elements.
But it is a little hard to understand.

Reply:
For me pointers look more clear. There is some memory overhead, but it's really low.

Reply:
OK

Reply:
More clear? Why? If so, would `uint32_t ***skipped_sdevs` be even more clear?

Sorry, I don't understand how come it can look more clear. Please explain.

Reply:
Rewritten as @huanghua78 suggested.